packages/browser/__tests__/Session.spec.ts

```diff
@@ -0,0 +1,205 @@
+        "handleIncomingRedirect"
+      );
+      const mySession = new Session({ clientAuthentication });
+      await mySession.handleIncomingRedirect("https://some.url");
+      expect(clientAuthnHandle).toHaveBeenCalled();
+    });
+
+    it("updates the session's info if relevant", async () => {
+      const clientAuthentication = mockClientAuthentication();
+      clientAuthentication.handleIncomingRedirect = jest.fn(
+        async (_url: string) => {
+          return {
+            isLoggedIn: true,
+            sessionId: "a session ID",
+            webId: "https://some.webid#them",
+          };
+        }
+      );
+      const mySession = new Session({ clientAuthentication });
+      expect(mySession.info.isLoggedIn).toEqual(false);
+      await mySession.handleIncomingRedirect("https://some.url");
+      expect(mySession.info.isLoggedIn).toEqual(true);
+      expect(mySession.info.sessionId).toEqual("a session ID");
+      expect(mySession.info.webId).toEqual("https://some.webid#them");
+    });
+
+    it("leaves the session's info unchanged if no session is obtained after redirect", async () => {
+      const clientAuthentication = mockClientAuthentication();
+      clientAuthentication.handleIncomingRedirect = jest.fn(
+        async (_url: string) => undefined
+      );
+      const mySession = new Session({ clientAuthentication }, "mySession");
+      await mySession.handleIncomingRedirect("https://some.url");
+      expect(mySession.info.isLoggedIn).toEqual(false);
+      expect(mySession.info.sessionId).toEqual("mySession");
+    });
+  });
+
+  describe("onLogin", () => {
+    it("calls the registered callback on login", async (done) => {
+      const myCallback = (): void => {
+        if (done) {
+          done();
+        }
+      };
+      const clientAuthentication = mockClientAuthentication();
+      clientAuthentication.handleIncomingRedirect = jest.fn(
+        async (_url: string) => {
+          return {
+            isLoggedIn: true,
+            sessionId: "a session ID",
+            webId: "https://some.webid#them",
+          };
+        }
+      );
+      const mySession = new Session({ clientAuthentication });
+      mySession.onLogin(myCallback);
+      await mySession.handleIncomingRedirect("https://some.url");
+    });
+
+    it("does not call the registered callback if login isn't successful", async () => {
+      const failCallback = (): void => {
+        fail();
+      };
+      const clientAuthentication = mockClientAuthentication();
+      clientAuthentication.handleIncomingRedirect = jest.fn(
+        async (_url: string) => {
+          return {
+            isLoggedIn: false,
+            sessionId: "a session ID",
+            webId: "https://some.webid#them",
+          };
+        }
+      );
+      const mySession = new Session({ clientAuthentication });
+      mySession.onLogin(failCallback);
+      await mySession.handleIncomingRedirect("https://some.url");
+    });
```

Comment:
Yeah, again just setting a simple boolean in the callback, and then asserting that it's **_not_** set here might be clearer...?

Reply:
If the purpose of the test is to check that the callback isn't called, isn't it even clearer to have the callback simpley `fail()` the test ?

Reply:
Yeah, arguably (so don't change if you prefer your approach). But personally I'd write this as:
```
it("does not call the registered callback if login isn't successful", async () => {
      const failCallback = (): void => {
        throw new Error("Should *NOT* call callback - this means test has failed!");
      };
      const clientAuthentication = mockClientAuthentication();
      clientAuthentication.handleIncomingRedirect = jest.fn(
        async (_url: string) => {
          return {
            isLoggedIn: false,
            sessionId: "a session ID",
            webId: "https://some.webid#them",
          };
        }
      );
      const mySession = new Session({ clientAuthentication });
      mySession.onLogin(failCallback);
      expect(() => await mySession.handleIncomingRedirect("https://some.url")).not.toThrow();
    });
```

Reply:
Actually, I thought that this could fail to capture the case where the callback is called after the promise resolves, but I could not produce such outcome, so I changed to your suggestion in 92e8ae0b14dc6a0577a72838a2cb4bf306fffe85.